### ISSUE TITLE
Unsupported DEPLOY-TOKEN header as authentication method

### ISSUE THREAD

**Issue by nyxgear** [python-gitlab/python-gitlab#1900]
Hi all, thanks for this very good python package! 

## Description of the problem, including code/CLI snippet

The [Gitlab object](https://github.com/python-gitlab/python-gitlab/blob/bd1ecdd5ad654b01b34e7a7a96821cc280b3ca67/gitlab/client.py#L61) does *not* support [deploy tokens](https://docs.gitlab.com/ee/user/project/deploy_tokens/) to authenticate with the GL APIs. I am aware of the [deploy token management](https://python-gitlab.readthedocs.io/en/stable/gl_objects/deploy_tokens.html#deploy-tokens) but that set of APIs is to *manage* deploy tokens, not to authenticate using them.

My use case: upload and download packages to GL package registry, using the [project.generic_packages.upload|download](https://python-gitlab.readthedocs.io/en/stable/gl_objects/packages.html#generic-packages) functions.

I don't want to use a `private_token` to authenticate since it is linked to a personal GL account and not to a project.

In GL documentation the DEPLOY-TOKEN header (for authentication) is not really well documented as instead it is the PRIVATE-TOKEN header. However, DEPLOY-TOKEN **works**, and in some sections like this [Gradle example](https://docs.gitlab.com/ee/user/packages/maven_repository/#authenticate-with-a-deploy-token-in-gradle
) it is indeed used.

I locally tested, live patching the Gitlab class,  how a deploy_token *could* work (I added DEPLOY-TOKEN header near [these lines](https://github.com/python-gitlab/python-gitlab/blob/bd1ecdd5ad654b01b34e7a7a96821cc280b3ca67/gitlab/client.py#L488))

```python
gl = gitlab.Gitlab(deploy_token="<deploy-token-here>")
project = gl.projects.get(12345, lazy=True) # lazy is mandatory if deploy_token is used, see explanation below
package = project.generic_packages.download(
    package_name="my-package",
    package_version="1.0.0",
    file_name="file.txt",
)
```

This example above works after my modifications.

### The need of "lazy" in `gl.projects.get()`

When you create a deploy token in a project, scopes you can pick are only:
- *read_repository*
- *read_package_registry*
- *write_package_registry*

The `gl.projects.get` would normally load the project attributes using the *api* or *read_api* scopes. This would result, if `lazy=True` is not used, in a 404 response. Indeed, since deploy tokens do not have this scope, `lazy=True` argument is required to skip preload of project attributes still allowing to, for example, upload download packages.

### Group tokens 

If the deploy_token auth method would be supported, then also [group deploy tokens](https://docs.gitlab.com/ee/user/project/deploy_tokens/#group-deploy-token) will then be supported. In fact, the HTTP header to be used is the same "DEPLOY-TOKEN" and not the "PRIVATE-TOKEN".

## Expected Behavior

Being able to authenticate to GitLab using "DEPLOY-TOKEN" header.

(I tested the use of currently supported "private_token" with a deploy token. It does not work.)

## Actual Behavior

The authentication method is not supported. 

## Specifications

  - python-gitlab version: **3.1.1**
  - API version you are using (v3/v4): **v4**
  - Gitlab server version (or gitlab.com): **gitlab.com** (current version: 14.9.0-pre 7bc6b3ba175)

----- 

I'd like to ask you a feedback and then, if you think it is worth, I'd like to submit a PR with the implementation. Thank you

**Comment by nejch** [comment]
My first thought would be that you should use project access tokens or group access tokens to avoid using tokens associated with your personal account. Those have the same granularity in scopes (https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#scopes-for-a-project-access-token) and work with the `Private-Token` header.

---

But specifically for deploy tokens - not sure there are use cases that project and group access tokens don't cover that deploy tokens do.. but since they exist, maybe we could support them :)

Note that deploy tokens are specifically [not designed for the API](https://docs.gitlab.com/ee/user/project/deploy_tokens/) (`Deploy tokens cannot be used with the GitLab API.`) and the reason we have this overlap is that we've decided to support the generic packages, which just happen to use the `api/v4` endpoint.

The caveat with requiring `lazy=True` should already be known to users due to limitations of the job token (if not, feel free to also improve the docs :)). So if you decide to open a PR please add a very visible warning admonition in the docs to warn users it is only usable with those very specific endpoints (and we will probably not support other package APIs as they do more than simple API calls).

**Comment by nyxgear** [comment]
Yes, you're right. The best for my case would be to use the project access tokens. I wasn't totally unaware of those.
However, by reading better the [docs](https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#scopes-for-a-project-access-token), I discovered they are a feature of GitLab Premium, not included in the free tier:
> On GitLab SaaS if you have the Premium license tier or higher. Project access tokens are not available with a [trial license](https://about.gitlab.com/free-trial/).

---
> Note that deploy tokens are specifically [not designed for the API](https://docs.gitlab.com/ee/user/project/deploy_tokens/) (Deploy tokens cannot be used with the GitLab API.) and the reason we have this overlap is that we've decided to support the generic packages, which just happen to use the api/v4 endpoint,

Uh! Now I see it. Sorry, I didn't see that particular sentence from the docs. Being now aware of that, I think that what I proposed with this issue totally loses meaning. Because of `Deploy tokens cannot be used with the GitLab API.` I honestly think that it has no sense to support them into python-gitlab package. At GitLab, they can decide that from tomorrow the deploy tokens won't be accepted in APIs auth anymore (since it is also so by docs) and everything will be broken. 

> I'm still a little uneasy about this as I think this will introduce more confusion that we already see with job tokens, but maybe there is value in it.

Indeed, I agree with you that supporting deploy tokens for auth will introduce more complexity and confusion for users of the package.

Just for completeness, here are the modifications to the `client.py` I did: https://github.com/nyxgear/python-gitlab/commit/c991b4b8bc2c8d28894f8940fc10671b50ff8b82

If you share my thoughts before we can just close the issue. Thanks for you fast and complete reply :)
